File: app/src/main/cpp/wn-steam-client/src/pb/cmsg_client_get_user_stats.cpp
#include "wn_steam/pb/cmsg_client_get_user_stats.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

std::vector<uint8_t> CMsgClientGetUserStats::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    // game_id and steam_id_for_user are fixed64 on the wire (NOT varint) —
    // see steammessages_clientserver_userstats.proto.
    w.fixed64_field(1, game_id);
    w.fixed64_field(4, steam_id_for_user);
    return out;
}

// Parse one CMsgClientGetUserStatsResponse.Achievement_Blocks sub-message.
//   1 uint32          achievement_id
//   2 repeated fixed32 unlock_time  — proto2 default is UNPACKED (each
//     element a separate field-2/Fixed32 tag), but a server that packs it
//     would send one field-2/LengthDelimited blob; handle both.
static std::optional<UserStatsAchievementBlock>
parse_achievement_block(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    UserStatsAchievementBlock blk;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        if (t->field_number == 1) {
            if (auto v = r.u32(); v) blk.achievement_id = *v;
            else return std::nullopt;
        } else if (t->field_number == 2 &&
                   t->wire_type == proto::WireType::Fixed32) {
            if (auto v = r.fixed32(); v) blk.unlock_time.push_back(*v);
            else return std::nullopt;
        } else if (t->field_number == 2 &&
                   t->wire_type == proto::WireType::LengthDelimited) {
            auto packed = r.bytes();
            if (!packed) return std::nullopt;
            proto::Reader pr(*packed);
            while (!pr.eof()) {
                auto v = pr.fixed32();
                if (!v) return std::nullopt;
                blk.unlock_time.push_back(*v);
            }
        } else if (!r.skip(t->wire_type)) {
            return std::nullopt;
        }
    }
    return blk;
}

std::optional<CMsgClientGetUserStatsResponse>
CMsgClientGetUserStatsResponse::deserialize(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CMsgClientGetUserStatsResponse m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 2:
                // int32 on the wire is a plain varint (two's complement).
                if (auto v = r.u64(); v)
                    m.eresult = static_cast<int32_t>(static_cast<uint32_t>(*v));
                else
                    return std::nullopt;
                break;
            case 3:
                if (auto v = r.u32(); v) m.crc_stats = *v;
                else return std::nullopt;
                break;
            case 4: {
                auto b = r.bytes();
                if (!b) return std::nullopt;
                m.schema.assign(b->begin(), b->end());
                break;
            }
            case 6: {
                auto b = r.bytes();
                if (!b) return std::nullopt;
                auto blk = parse_achievement_block(*b);
                if (!blk) return std::nullopt;
                m.achievement_blocks.push_back(std::move(*blk));
                break;
            }
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cmsg_client_get_user_stats.rs
use crate::proto_wire::{Reader, WireType, Writer};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct CMsgClientGetUserStats {
    pub game_id: u64,
    pub steam_id_for_user: u64,
}

impl CMsgClientGetUserStats {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.fixed64_field(1, self.game_id);
        w.fixed64_field(4, self.steam_id_for_user);
        out
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct UserStatsAchievementBlock {
    pub achievement_id: u32,
    pub unlock_time: Vec<u32>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CMsgClientGetUserStatsResponse {
    pub eresult: i32,
    pub crc_stats: u32,
    pub schema: Vec<u8>,
    pub achievement_blocks: Vec<UserStatsAchievementBlock>,
}

impl Default for CMsgClientGetUserStatsResponse {
    fn default() -> Self {
        Self {
            eresult: 2,
            crc_stats: 0,
            schema: Vec::new(),
            achievement_blocks: Vec::new(),
        }
    }
}

impl CMsgClientGetUserStatsResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                2 => msg.eresult = reader.u64()? as u32 as i32,
                3 => msg.crc_stats = reader.u32()?,
                4 => msg.schema = reader.bytes()?.to_vec(),
                6 => msg
                    .achievement_blocks
                    .push(parse_achievement_block(reader.bytes()?)?),
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

fn parse_achievement_block(body: &[u8]) -> Option<UserStatsAchievementBlock> {
    let mut reader = Reader::new(body);
    let mut block = UserStatsAchievementBlock::default();
    while !reader.eof() {
        let Some(tag) = reader.next_tag() else {
            return reader.ok().then_some(block);
        };
        match (tag.field_number, tag.wire_type) {
            (1, _) => block.achievement_id = reader.u32()?,
            (2, WireType::Fixed32) => block.unlock_time.push(reader.fixed32()?),
            (2, WireType::LengthDelimited) => {
                let mut packed = Reader::new(reader.bytes()?);
                while !packed.eof() {
                    block.unlock_time.push(packed.fixed32()?);
                }
            }
            _ => {
                if !reader.skip(tag.wire_type) {
                    return None;
                }
            }
        }
    }
    Some(block)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_unpacked_and_packed_achievement_unlocks() {
        let mut block = Vec::new();
        {
            let mut w = Writer::new(&mut block);
            w.uint32_field(1, 32);
            w.tag(2, WireType::Fixed32);
            w.raw_bytes(&10u32.to_le_bytes());
            let mut packed = Vec::new();
            packed.extend_from_slice(&20u32.to_le_bytes());
            packed.extend_from_slice(&30u32.to_le_bytes());
            w.bytes_field(2, &packed);
        }

        let mut body = Vec::new();
        {
            let mut w = Writer::new(&mut body);
            w.int32_field(2, 1);
            w.uint32_field(3, 1234);
            w.bytes_field(4, b"schema");
            w.submessage_field(6, &block);
        }

        let parsed = CMsgClientGetUserStatsResponse::deserialize(&body).unwrap();
        assert_eq!(parsed.eresult, 1);
        assert_eq!(parsed.crc_stats, 1234);
        assert_eq!(parsed.schema, b"schema");
        assert_eq!(parsed.achievement_blocks[0].unlock_time, vec![10, 20, 30]);
    }
}
